### PR TITLE
record_function: update to use custom_class API

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #72303
* __->__ #76420

Re-submit of gh-72302

This still has a small performance hit, but it much smaller. On my
machine I see `_record_fucntion_exit._RecordFunction` takes 1.05 us
compared to the `Tensor` overload taking 0.79 us.

In an overall comparison, I see a 0.7 us slowdown from 6.0 us to
6.7 us for this timeit benchmark
```python
import torch

def foo():
  with torch.profiler.record_function("foo"):
    return torch.eye(3)

%timeit foo()
```

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @EikanWang @jgong5 @wenzhe-nrv @sanchitintel

### CODE REVIEW

File: test/test_profiler_tree.py [hunk 1]
```diff
@@ -316,34 +316,18 @@ def test_profiler_experimental_tree_with_record_function(self):
         self.assertTreesMatch(
             ProfilerTree.format(p.profiler, 12),
             """\
-            aten::zeros
-              aten::empty
-              aten::zero_
```
Comment (peterbell10) [inline test/test_profiler_tree.py:321]: This is expected because the `aten::zeros` call was coming from `record_function.__init__`.

**Comment by Lezcano** [comment]
@robieta could you take al look?

**Review by robieta — APPROVED** [review]
LGTM. Sorry this slipped under my radar.

**Pull request closed without merging**